ibis/expr/analysis.py

```diff
@@ -774,28 +775,13 @@ def predicate(node):
 
 def find_subqueries(node: ops.Node) -> Counter:
     counts = Counter()
-
-    def finder(node: ops.Node):
-        if isinstance(node, ops.Join):
-            return [node.left, node.right], None
-        elif isinstance(node, ops.PhysicalTable):
-            return g.halt, None
-        elif isinstance(node, ops.SelfReference):
-            return g.proceed, None
-        elif isinstance(node, (ops.Selection, ops.Aggregation)):
-            counts[node] += 1
-            return [node.table], None
-        elif isinstance(node, ops.TableNode):
-            counts[node] += 1
-            return g.proceed, None
-        elif isinstance(node, ops.TableColumn):
-            return node.table not in counts, None
-        else:
-            return g.proceed, None
-
-    # keep duplicates so we can determine where an expression is used
-    # more than once
-    list(g.traverse(finder, node, dedup=False))
+    for graph in map(g.Graph.from_dfs, util.promote_list(node)):
```

Comment:
This code segment is pretty hard to interpret

Reply:
Which part?

Here's what's happening:

`node` can be either a `Node` or a sequence of `Node`s, so I use `promote_list` to always make it a list of nodes for ease of processing and code clarity.

Then I convert each node in the list to a `Graph` and compute the number of unique table **dependents** of the current node, but only if the current node is a selection/aggregation. Selection/aggregation maps to a subquery in SQL.

The number of unique dependents of the current node is the number of unique usages of the current subquery, which is used subsequently in the compiler to determine whether an expression should be compiled to a CTE.

Reply:
This isn't that different from the existing implementation. The previous implementation was a bit less aggressive in counting subqueries.